README.md
# Page-Solutions #

The Page-Solutions plugin adds enhanced widget areas and page caching functionality.

## Description ##

The Page-Solutions plugin provides custom CSS and JavaScript modifications, virtual widget mapping and page caching functionality on a page by page basis. 
This efficient and powerful plugin is well suited for page-intensive and non-blog WordPress applications. 
The Page-Solutions plugin is developed and maintained by <a href="https://www.usi2solve.com">Universal Solutions</a>.

## Page Caching ##
The Page-Solutions plugin stores content in the database for quick access which improves performance by eliminating the overhead of loading and running WordPress for pages that have not changed recently. In order to see changes however, the page cache must be cleared whenever you edit a page or if a page is updated by a widget running in one the theme's widget areas. The following four options allow you to control the page cache:

   * **Inherit parent page cache settings** - The cache settings are inherited from the parent page. Check this feature if this is a child page and its layout and function is similar to it's parent's page.

   * **Allow widgets to clear cache** - Allow widget(s) in your theme's widget area to dynamically clear the cache as the widget(s) desire. Check this feature if you use widget(s) that were designed to use the Page-Solutions caching system.

   * **Clear cache on next update** - The cache is cleared the next time the Update button is clicked. Check this feature if you don't want to remember to clear the cash manualy the next time you update your changes.

   * **Clear cache on every update** - The cache is cleared every time the Update button is clicked. Check this feature if you don't want to remember to clear the cash manualy every time you update your changes.

The following four options allow you to control how the page cache is cleared.

   * **Disable cache** - The cache is not used for the current page. Select this option if you don't want to use the caching features for this page or if the page is very dynamic and can rarely be re-used. This is the default option.

   * **Clear cache manually** - You manually clear the cache after you edit the page. Select this option if this page is only changed by you and content is never changed by widget. Make sure you click the Clear Cache button when you finish your page edits or your changes will not be seen by the world.

   * **Clear cache every** - The cache is cleared after the given time period has expired. Select this option if page content is changed by a widget(s) but it's not necessary for the changes to show immediately. Specify the period with the drop down box under this option.

   * **Clear cache everyday at** - The cache is cleared based on the given schedule. Select this option if page content is changed by a widget(s) and you want to ensure that changes are show at specific times of the day. List the times when the cache should be cleared under this option.

The Page-Solutions cache features and options are configured on a page by page basis.

## Page Options ##
The Page-Solutions plugin adds the following page options:

   * **Accepts arguments** - The page can be called with an argument string that follows the page URL. This argument string can be used to pass information into widgets that have been designed to use this feature. This option is disabled if this page is a parent page.

The Page-Solutions plugin options are configured on a page by page basis.

## Installation ##
The Page-Solutions plugin follows the standard WordPress <a href="https://codex.wordpress.org/Managing_Plugins#Manual_Plugin_Installation">manual plugin installation</a> procedure:
1. Clone or Download the Page-Solutions archive to your computer.
1. Extract the archive contents to your local file system.
1. Rename the extracted folder to ` usi-page-solutions ` if not already done so during the extraction.
1. Upload the ` .php ` files from your ` usi-page-solutions ` folder to the ` wp-content/plugins ` folder in your target WordPress installation.
1. Activate the plugin via the WordPress *Plugins* menu located on the left side bar.

## License ##
> Page-Solutions is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License 
as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

> Page-Solutions is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

> You should have received a copy of the GNU General Public License along with Page-Solutions.  If not, see 
<http://www.gnu.org/licenses/>.

## Donations ##
Donations are accepted at <a href="https://www.usi2solve.com/donate/page-solutions">www.usi2solve.com/donate</a>. Thank you for your support!

CHANGELOG.md
# Change Log #

Page-Solutions plugin changes are logged here using <a href="http://semver.org/">Semantic Versioning</a>.

## 1.7.3 (2023-08-13) ##
* Removed GitHub update functionality.

## 1.7.2 (2023-03-16) ##
* Fixed CSS page layout edit feature.

## 1.7.1 (2023-01-25) ##
* Added global header content functionality.

## 1.7.0 (2022-08-09) ##
* Set all versions numbers to same version.
* Improved debug handling.
* Changed the capture method to not use output buffering.

## 1.6.0 (2021-06-12) ##
* Set all versions numbers to same version.
* Improved logging and database connection functions.

## 1.5.6 (2020-03-31) ##
* Added use of 'visual-grid' option and 'action_admin_head' convenience function.

## 1.5.5 (2020-03-24) ##
* Updated settings class argument.

## 1.5.4 (2020-02-28) ##
* Updated settings handling.

## 1.5.3 (2020-02-20) ##
* Updated capability and updates handling, set all versions numbers to same version.

## 1.5.2 (2020-02-06) ##
* Updated all plugin version numbers to match.

## 1.5.1 (2020-02-05) ##
* Fixed static class reference error.

## 1.5.0 (2020-01-12) ##
* Added license and copyright notice.

## 1.4.1 (2020-01-12) ##
* Added 'Updates' tab to settings page to optionally allow GitHub updates.
* Added license and copyright notice.
* Changed format of version number on main plugin file header to facilitate Git Hub update.
* Updated all plugin version numbers to match.

## 1.4.0 (2019-11-27) ##
* Removed usi-dbs module and replaced with usi-page-dbs module.
* Updated all plugin version numbers to match.

## 1.3.0 (2019-06-15) ##
* Removed usi-settings files from sub-folder under parent plugin and made a stand alone class.
* Updated all plugin version numbers to match.

## 1.2.1 (2018-10-07) ##
* Updated all class versions to same version number.

## 1.2.0 (2018-09-04) ##
* Added clear all page cache information in settings. This is useful when a change is made that globally effects the site, like a menu change or template modification.

## 1.1.1 (2018-01-17) ##
* Updated page caching documentation.
* Updated page options documentation.

## 1.1.0 (2018-01-13) ##
* Updated all class versions to same version number.
* Added script and style links to layout functionality.
* Moved database and debugging files into the usi-library folder and made them a Git submodule.

## 1.0.1 (2018-01-10) ##
* Moved usi-settings files into their own folder and made them a Git submodule.

## 1.0.0 (2018-01-07) ##
* Reduced length of enhanced widgets option name.
* Add ability to surpress the settings link added to the installed plugins page.
* Added mysqli database support functions to repository.
* Updated all class versions to same version.
* Converted code and CSS page editors to use settings API.
* Improved recusive saving of page options and parameters.

## 0.0.3 (2018-01-05) ##
* Improved and standardized debug logging.
* Replaced meta value access functions with meta_value_get() and meta_value_put().
* Save page meta value as base64 string to address serialization errors.
* Changed name of edit virtual list permission.

## 0.0.2 (2018-01-04) ##
* Simplified some PHP statements for assignments done in "empty()" function.

## 0.0.1 (2018-01-03) ##
* Started adding source files to GIT.

## 0.0.0 (2017-11-11) ##
* Initial release.



usi-page-solutions.php
<?php // ------------------------------------------------------------------------------------------------------------------------ //

defined('ABSPATH') or die('Accesss not allowed.');

/* 
Author:            Jim Schwanda
Author URI:        https://www.usi2solve.com/leader
Copyright:         2023 by Jim Schwanda.
Description:       The Page-Solutions plugin provides custom CSS and JavaScript modifications, virtual widget mapping and page caching functionality on a page by page basis. This efficient and powerful plugin is well suited for page-intensive and non-blog WordPress applications. The Page-Solutions plugin is developed and maintained by Universal Solutions.
Donate link:       https://www.usi2solve.com/donate/wordpress-solutions
License:           GPL-3.0
License URI:       https://github.com/jaschwanda/wordpress-solutions/blob/master/LICENSE.md
Plugin Name:       Page-Solutions
Plugin URI:        https://github.com/jaschwanda/page-solutions
Requires at least: 5.0
Requires PHP:      5.6.25
Tested up to:      7.4.19
Text Domain:       usi-page-solutions
Version:           1.7.2
Warranty:          This software is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
*/

// remove validate() from this file and use template or index-cache.php
// change debug to use logging functions;
// make sure changing parent pages options aren't propagated down on childeren not inheriting;

final class USI_Page_Solutions {

   const VERSION = '1.7.2 (2023-03-16)';

   const DATE_ALPHA   = '0000-00-00 00:00:00'; // remove with validate();
   const DATE_OMEGA   = '9999-12-31 23:59:59'; // put in template;

   const DEBUG_OFF    = 0x00000000;
   const DEBUG_CACHE  = 0x13000001;
   const DEBUG_HTML   = 0x13000002;
   const DEBUG_SANITZ = 0x13000004;

   const NAME         = 'Page-Solutions';
   const POST_META    = '_usi-page-solutions';
   const PREFIX       = 'usi-page';
   const TEXTDOMAIN   = 'usi-page-solutions';

   const WIDGETS_INIT_PRIORITY = 100;

   public static $capabilities = array(
      'enhanced-edit'  => 'Select enhanced widget areas|administrator',
      'virtual-add'    => 'Add virtual widget collections|administrator',
      'virtual-edit'   => 'Edit virtual widget collections|administrator',
      'virtual-delete' => 'Delete virtual widget collections|administrator',
      'settings-view'  => 'View settings page|administrator',
   );

   private static $debug = self::DEBUG_OFF;
   private static $info  = null;
   private static $theme = null;

   public static $meta_value = null; // Page/post postmeta data;
   public static $option_name_base = null;
   public static $option_name_virtual = null;
   public static $options = array();
   public static $options_virtual = null; // Virtual widget array for current theme;
   public static $theme_name = null; // Slugified name of current theme;
   public static $virtual_source = null; // Virtual source widget to map in enhanced widget area;

   private function __construct() {
   } // __construct();

   static function action_admin_notices() {
      global $pagenow;
      if ('plugins.php' == $pagenow) {
        $text = sprintf(
           __('The %s plugin is required for the %s plugin to run properly.', self::TEXTDOMAIN), 
           '<b>WordPress-Solutions</b>',
           '<b>Page-Solutions</b>'
        );
        echo '<div class="notice notice-warning is-dismissible"><p>' . $text . '</p></div>';
      }
   } // action_admin_notices();

   static function action_widgets_init() {
      $theme = wp_get_theme();
      self::$theme_name = sanitize_title($theme->get('Name'));
      // self::$option_name_base = 'usi-page-solutions-options-' . self::$theme_name; // This seems too f++king long;
      self::$option_name_base = 'usi-page-' . self::$theme_name;
      self::$option_name_virtual = self::$option_name_base . '-virtual';
      self::$options_virtual = get_option(self::$option_name_virtual);
      if (!empty(self::$options_virtual)) {
         for ($ith = 0; $ith < count(self::$options_virtual); $ith++) {
            register_sidebar(self::$options_virtual[$ith]);
         }
      }
   } // action_widgets_init();

   static function action_wp_body_open() {
      echo USI_Page_Solutions::$options['preferences']['global-header'];
   } // action_wp_body_open();

   static function action_wp_enqueue_scripts() {

      global $post;

      $post_id = (int)(!empty($post->ID) ? $post->ID : 0);

      self::$meta_value = self::meta_value_get($post_id);

      $what2do = array(
         array('key' => 'styles_parent',  'inherit' => 'styles_inherit',  'funtction' => 'wp_enqueue_style',  'default' => null),
         array('key' => 'styles',         'inherit' => false,             'funtction' => 'wp_enqueue_style',  'default' => null),
         array('key' => 'scripts_parent', 'inherit' => 'scripts_inherit', 'funtction' => 'wp_enqueue_script', 'default' => false),
         array('key' => 'scripts',        'inherit' => false,             'funtction' => 'wp_enqueue_script', 'default' => false),
      );

      foreach ($what2do as $do) {
         if (!$do['inherit'] || !empty(self::$meta_value['options'][$do['inherit']])) {
            if (!empty(self::$meta_value['layout'][$do['key']])) {
               $link = self::$meta_value['layout'][$do['key']];
               foreach ($link as $key => $value) {
                  if (!empty($value)) {
                     $tokens = explode(' ', $value);
                     $do['funtction'](
                        $tokens[0], // Unique link sheet name;
                        $tokens[1], // Full URL of the link sheet, or path of link relative to root;
                        null, // Array of registered link handles this link depends on;
                        (!empty($tokens[2]) && ('null' != $tokens[2])) ? $tokens[2] : null, // Link version number;
                        !empty($tokens[3]) ? $tokens[3] : $do['default'] // Media flag;
                     );
                  }
               }
            }
         }
      }

   } // action_wp_enqueue_scripts();

   static function action_wp_footer() {
      if (!empty(self::$meta_value['layout']['codes_foot_parent'])) echo self::$meta_value['layout']['codes_foot_parent'];
      if (!empty(self::$meta_value['layout']['codes_foot'])) echo self::$meta_value['layout']['codes_foot'];
   } // action_wp_footer();

   static function action_wp_head() {
      if (self::$meta_value['options']['css_inherit'] && !empty(self::$meta_value['layout']['css_parent'])) echo self::$meta_value['layout']['css_parent'];
      if (!empty(self::$meta_value['layout']['css'])) echo self::$meta_value['layout']['css'];
      if (!empty(self::$meta_value['layout']['codes_head_parent'])) echo self::$meta_value['layout']['codes_head_parent'];
      if (!empty(self::$meta_value['layout']['codes_head'])) echo self::$meta_value['layout']['codes_head'];
   } // action_wp_head();

   static function callback() {
    
      global $wp_registered_widgets;

      self::$info = array();

      $original_params = func_get_args();
      $widget_id = $original_params[0]['widget_id'];   
      $wp_registered_widgets[$widget_id]['callback'] = $original_callback = $wp_registered_widgets[$widget_id]['original_callback'];
    
      if (is_callable($original_callback)) {
         self::$info['class'] = get_class($original_callback[0]);
         if (method_exists($original_callback[0], 'USI_Page_Cache_file')) {
            $file = $original_callback[0]->USI_Page_Cache_file();
         } else {
            $rf = new ReflectionClass(self::$info['class']);
            $file = $rf->getFileName();
         }
         self::$info['file'] = str_replace(str_replace('/', '\\', $_SERVER['DOCUMENT_ROOT']), '', $file);
         self::$info['begin'] = ob_get_length();
         call_user_func_array($original_callback, $original_params);
         self::$info['end'] = ob_get_length();
         if (!empty(self::$info['instance']['is_dynamic'])) USI_Page_Cache::que(self::$info);

      }
    
   } // callback();

   static function filter_dynamic_sidebar_params($sidebar_params) {
      // Override theme html with virtual widget values;
      static $ps_offset = 0;
      if (!empty(self::$virtual_source[$ps_offset])) {
         $virtual_widget_id = self::$virtual_source[$ps_offset++];
         if (!empty(self::$options_virtual)) {
            for ($ith = 0; $ith < count(self::$options_virtual); $ith++) {
               if ($virtual_widget_id == self::$options_virtual[$ith]['id']) {
                  $option_virtual = self::$options_virtual[$ith];
                  if ('disable' != $option_virtual['before_title'])  $sidebar_params[0]['before_title']  = $option_virtual['before_title'];
                  if ('disable' != $option_virtual['after_title'])   $sidebar_params[0]['after_title']   = $option_virtual['after_title'];
                  if ('disable' != $option_virtual['before_widget']) $sidebar_params[0]['before_widget'] = $option_virtual['before_widget'];
                  if ('disable' != $option_virtual['after_widget'])  $sidebar_params[0]['after_widget']  = $option_virtual['after_widget'];
                  break;
               }
            }
         }
      }
      
      // This filter intercepts the widget() calls so that the widget output can be buffered and
      // the dynamics can be saved, not required when caching is disabled as everything must run;
      if (is_admin() || ('disable' == self::$meta_value['cache']['mode'])) return($sidebar_params);

      global $wp_registered_widgets;

      $widget_id = $sidebar_params[0]['widget_id'];
 
      $wp_registered_widgets[$widget_id]['original_callback'] = $wp_registered_widgets[$widget_id]['callback'];
      $wp_registered_widgets[$widget_id]['callback'] = array(__CLASS__, 'callback');
      return($sidebar_params);

   } // filter_dynamic_sidebar_params();

   static function filter_pre_option_stylesheet() { 

      return(self::load_theme(1)); 

   } // filter_pre_option_stylesheet();

   static function filter_pre_option_template() { 

      return(self::load_theme(0)); 

   } // filter_pre_option_template();

   static function filter_sidebars_widgets($sidebars_widgets) {
   
      static $mapped_widgets = null;
   
      if (is_page()) {
   
         if ($mapped_widgets) return($mapped_widgets);

         if (!empty(self::$meta_value['widgets'])) {
            foreach (self::$meta_value['widgets'] as $target_id => $enhanced_widget_areas) {
               $jth = 0;
               foreach ($enhanced_widget_areas as $source_id) {
                  for ($ith = 0; $ith < count($sidebars_widgets[$source_id]); $ith++, $jth++) {
                     $sidebars_widgets[$target_id][$jth] = $sidebars_widgets[$source_id][$ith];
                     self::$virtual_source[] = $source_id;
                  }
               }
            }
         }
         $mapped_widgets = $sidebars_widgets;
      }

      return($sidebars_widgets);
   
   } // filter_sidebars_widgets();

   static function filter_widget_display_callback($instance, $that, $args) {
      self::$info['args'] = $args;
      return(self::$info['instance'] = $instance);
   } // filter_widget_display_callback();

   static function init() {

      if (empty(USI_Page_Solutions::$options)) {
         $defaults['cache']['config-location'] =
         $defaults['cache']['root-location']   =
         $defaults['cache']['debug-ip']        = '';
         $defaults['cache']['track-times']     =
         $defaults['cache']['debug-meta-data'] = 
         $defaults['cache']['debug-sql']       = 
         $defaults['preferences']['enable-cache'] = 
         $defaults['preferences']['enable-enhanced-areas'] = 
         $defaults['preferences']['enable-layout'] = false;
         $defaults['preferences']['global-header'] = null;
         $defaults['preferences']['page-mru-max'] = 
         $defaults['preferences']['post-mru-max'] = 4;
         USI_Page_Solutions::$options = get_option(self::PREFIX . '-options', $defaults);
         if (!is_admin()) {
            if (empty(USI_Page_Solutions::$options['cache']['root-location'])) {
               $included_files = get_included_files();
               USI_Page_Solutions::$options['cache']['root-location'] = $included_files[0];
               update_option(self::PREFIX . '-options', USI_Page_Solutions::$options);
            }
         }
      }

      if (!empty(USI_Page_Solutions::$options['preferences']['enable-enhanced-areas'])) {
         add_action('widgets_init', array(__CLASS__, 'action_widgets_init'), self::WIDGETS_INIT_PRIORITY);
         add_filter('dynamic_sidebar_params', array(__CLASS__, 'filter_dynamic_sidebar_params'));
         add_filter('sidebars_widgets', array(__CLASS__, 'filter_sidebars_widgets'));
         add_filter('widget_display_callback', array(__CLASS__, 'filter_widget_display_callback'), 10, 3);
      }

      if (!empty(USI_Page_Solutions::$options['preferences']['enable-layout'])) {
         add_action('wp_enqueue_scripts', array(__CLASS__, 'action_wp_enqueue_scripts'), 20);
         add_action('wp_footer', array(__CLASS__, 'action_wp_footer'), 9999);
         add_action('wp_head', array(__CLASS__, 'action_wp_head'), 100);
      }

      if (!empty(USI_Page_Solutions::$options['preferences']['global-header'])) {
         add_action('wp_body_open', array(__CLASS__, 'action_wp_body_open'));
      }
/*
      if (!empty(USI_Page_Cache::$theme) && ('default' != USI_Page_Cache::$theme)) {
         self::$theme = explode(':', USI_Page_Cache::$theme);
         add_filter('pre_option_stylesheet', [__CLASS__, 'filter_pre_option_stylesheet']);
         add_filter('pre_option_template',   [__CLASS__, 'filter_pre_option_template']);
      }
*/
   } // init();

   static public function load_theme($index) {
      return(self::$theme[$index] ?? '');
   } // load_theme();

   static function meta_value_get($post_id, $debug = false) {

      $data = [];
      if ($post_id) {
         $results = get_post_meta($post_id);
         if (!empty($results)) {
            foreach ($results as $index => $value) {
               if (str_starts_with( $index, self::POST_META)) {
                  $data = unserialize(base64_decode($value[0]));
               }
            }
         }
      }

      $layout['codes_foot']          = $data['layout']['codes_foot']        ?? null;
      $layout['codes_foot_parent']   = $data['layout']['codes_foot_parent'] ?? null;
      $layout['codes_head']          = $data['layout']['codes_head']        ?? null;
      $layout['codes_head_parent']   = $data['layout']['codes_head_parent'] ?? null;
      $layout['css']                 = $data['layout']['css']               ?? null;
      $layout['css_parent']          = $data['layout']['css_parent']        ?? null;
      $layout['scripts']             = $data['layout']['scripts']           ?? null;
      $layout['scripts_parent']      = $data['layout']['scripts_parent']    ?? null;
      $layout['styles']              = $data['layout']['styles']            ?? null;
      $layout['styles_parent']       = $data['layout']['styles_parent']     ?? null;

      $options['arguments']          = !empty($data['options']['arguments']);
      $options['codes_foot_inherit'] = !empty($data['options']['codes_foot_inherit']);
      $options['codes_head_inherit'] = !empty($data['options']['codes_head_inherit']);
      $options['css_inherit']        = !empty($data['options']['css_inherit']);
      $options['scripts_inherit']    = !empty($data['options']['scripts_inherit']);
      $options['styles_inherit']     = !empty($data['options']['styles_inherit']);
      $options['theme']              = $data['options']['theme'] ?? 'default';
      $options['widgets_inherit']    = !empty($data['options']['widgets_inherit']);

      $widgets = $data['widgets'] ?? null;

      $url      = rtrim(get_permalink($post_id), '/') . '/';
      $path     = str_replace(array($_SERVER['SERVER_NAME'], 'https://', 'http://'), '', $url);
      $meta_key = self::POST_META . ($options['arguments'] ? '*' : '!') . $path;

      $value = [
         'key'     => $meta_key, 
         'post_id' => $post_id, 
         'cache'   => null, 
         'layout'  => $layout, 
         'options' => $options, 
         'widgets' => $widgets
      ];

      if (!empty(USI_Page_Solutions::$options['preferences']['enable-cache'])) {
         $value['cache'] = self::validate(!empty($data['cache']) ? $data['cache'] : null);
      }

      if (self::$debug || $debug) {

         if (!(self::$debug & self::DEBUG_HTML)) { 
            $html = $value['cache']['html']; 
            $value['cache']['html'] = '~~~'; 
         }

         usi::log2(__METHOD__.'()~'.__LINE__.':$value=', $value);

         if (!(self::$debug & self::DEBUG_HTML)) {
            $value['cache']['html'] = $html;
         }

      }

      return($value);

   } // meta_value_get();

   static function meta_value_put($value, $debug = false) {

      if (self::$debug || $debug) {

         if (!(self::$debug & self::DEBUG_HTML)) { 
            $html = $value['cache']['html']; 
            $value['cache']['html'] = '~~~'; 
         }

         usi::log2(__METHOD__.'()~'.__LINE__.':$value=', $value);

         if (!(self::$debug & self::DEBUG_HTML)) {
            $value['cache']['html'] = $html;
         }

      }

      update_post_meta($value['post_id'], $value['key'], base64_encode(serialize($value)));

   } // meta_value_put();

   public static function number_of_offspring($page_id) {
      global $wpdb;
      $SAFE_table_name = $wpdb->prefix . 'posts';
      $children = (int)$wpdb->get_var(
         $wpdb->prepare("SELECT count(`ID`) FROM `$SAFE_table_name` WHERE (`post_parent` = %d) AND (`post_type` = 'page')", $page_id)
      );
      return($children);
   } // number_of_offspring

   // remove DATE_ALPHA
   private static function validate($cache) { 
      $allow_clear = !empty($cache['allow-clear']);
      $clear_every_publish = !empty($cache['clear-every-publish']);
      $inherit_parent = !empty($cache['inherit-parent']);
      $dynamics = isset($cache['dynamics']) ? $cache['dynamics'] : [];
      switch ($mode = isset($cache['mode']) ? $cache['mode'] : 'disable') {
      default: $mode = 'disable'; case 'manual': case 'period': case 'schedule': break;
      }
      switch ($period = (int)(isset($cache['period']) ? $cache['period'] : 86400)) {
      default: $period = 86400;
      case 300:   case 600:   case 900:   case 1200:  case 1800:  case 3600:  case 7200:  
      case 10800: case 14400: case 21600: case 28800: case 43200: break;
      }
      $schedule = isset($cache['schedule']) ? $cache['schedule'] : ['00:00:00'];
      $safe_schedule = [];
      foreach ($schedule as $time) {
         if (preg_match('/([0-1][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9])/', $time)) $safe_schedule[] = $time;
      }
      if (empty($safe_schedule)) { $safe_schedule = ['00:00:00']; } else { sort($safe_schedule); }
      $updated     = isset($cache['updated']) && preg_match('/([0-1][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9])/', $cache['updated']) ? $cache['updated'] : self::DATE_ALPHA;
      $valid_until = isset($cache['valid_until']) &&  preg_match('/([0-1][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9])/', $cache['valid_until']) ? $cache['valid_until'] : self::DATE_ALPHA;
      if (isset($cache['html'])) {
         $html     = $cache['html'];
      } else {
         $html     = '';
         $updated  = $valid_until = self::DATE_ALPHA;
      }
      $size = strlen($html);
      return(['allow-clear' => $allow_clear, 'clear-every-publish' => $clear_every_publish, 'inherit-parent' => $inherit_parent, 
         'mode' => $mode, 'period' => $period, 'schedule' => $safe_schedule, 'size' => $size, 'updated' => $updated, 
         'valid_until' => $valid_until, 'dynamics' => $dynamics, 'html' => $html]);
   } // validate();

} // Class USI_Page_Solutions;

USI_Page_Solutions::init();

if (is_admin() && !defined('WP_UNINSTALL_PLUGIN')) {
   if (is_dir(plugin_dir_path(__DIR__) . 'usi-wordpress-solutions')) {
      require_once('usi-page-solutions-admin.php');
      require_once('usi-page-solutions-install.php');
      require_once('usi-page-solutions-layout.php');
      require_once('usi-page-solutions-layout-edit.php');
      require_once('usi-page-solutions-settings.php');
      require_once('usi-page-solutions-virtual.php');
      require_once('usi-page-solutions-virtual-list.php');
      if (!empty(USI_Page_Solutions::$options['preferences']['enable-cache'])) {
         require_once('usi-page-solutions-cache.php');
         require_once('usi-page-solutions-options.php');
      }
   } else {
      add_action('admin_notices', array('USI_Page_Solutions', 'action_admin_notices'));
   }
}

// --------------------------------------------------------------------------------------------------------------------------- // ?>